Kernel Schedule

sergisiso opened the pull request:
Add functionality to PSyAlg `Kern` calls (method `get_kernel_schedule()`) to provide the representation of the targeted kernel. This constructs a `KernelSchedule` with a Symbol table and the Execution part of the code.

sergisiso commented:
Ready for first review @rupertford @arporter

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
+    symbol table.
+    '''
+    def __init__(self):
+        self._symbols = {}
```
Please add a comment describing what `self._symbols` will hold. e.g. "A dict of Symbol objects with the symbol names as keys."

sergisiso replied:
Done

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
+
+    def lookup(self, name):
+        '''
+        Look up a symbol in the symbols table.
```
s/symbols/symbol/

sergisiso replied:
Done

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
+        '''
+        Print a representation of this Symbol Table to stdout.
+        '''
+        print("Symbol Table:")
```
Could you do `print(str(self))` instead and just make use of the `__str__` method?

sergisiso replied:
I did it because I think it will often be used in conjunction with the schedule view, and from the 2 scripts below the former looks nicer:
```
kernel_code = kernel.get_kernel_schedule()
kernel_code.symbol_table.view()
kernel_code.view()
```
```
kernel_code = kernel.get_kernel_schedule()
print(str(kernel_code.symbol_table))
kernel_code.view()
```
But it is true than functionally they are the same. I am happy with either option.

arporter replied:
I think what I meant was that the view() method could just print the results of the str() method:

    def view(self):
        print(str(self))

Does that work?

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
+        Print a text representation of this node to stdout and then
+        call the view() method of any children.
+
+        :param indent: Depth of indent for output text
```
Again, this could just be `:param int indent: xxxx`

sergisiso replied:
Done

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
+            entity.view(indent=indent + 1)
+
+    def __str__(self):
+        result = "Schedule[name:'" + self._name + "']:\n"
```
I'm not sure whether we should be outputting "KernelSchedule" instead of "Schedule". Maybe by the end of this review I'll have decided :-)

arporter replied:
I've decided I'm happy that it's still a Schedule.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -526,7 +527,7 @@ def test_derived_type_deref_naming():
     assert output in generated_code
 
-
+# TODO: Update considering fparser/166 issue resolution.
```
I think you can safely remove this TODO as fparser is doing the right thing.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2617,8 +2627,269 @@ def test_binaryoperation_can_be_printed():
     assert "BinaryOperation[operator:'+']\n" in str(binaryOp)
 
 
+# Test KernelSchedule Class
```
Please add docstrings for the new tests.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2617,8 +2627,269 @@ def test_binaryoperation_can_be_printed():
+    assignment.addchild(rhs)
+    kschedule.addchild(assignment)
+    assert "Schedule[name:'kname']:\n" in str(kschedule)
+    assert "Assignment" in str(kschedule)  # Check childs are printed
```
s/childs/children/

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2617,8 +2627,269 @@ def test_binaryoperation_can_be_printed():
+    schedule = processor.generate_schedule("dummy_code", ast2)
+    assert isinstance(schedule, KernelSchedule)
+
+    # Test generate an unexistent subroutine name
```
s/Test generate an unexistant/Test that we get an error for a nonexistant/

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2617,8 +2627,269 @@ def test_binaryoperation_can_be_printed():
+        schedule = processor.generate_schedule("dummy_code", ast2)
+    assert "Unexpected kernel AST. Could not find specification part."
+
+    DUMMY_KERNEL_METADATA = '''
```
I think you should break the rest of this test out into a separate one as it deals with a non-empty kernel.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2617,8 +2627,269 @@ def test_binaryoperation_can_be_printed():
+
+    # Test simple declarations
+    reader = FortranStringReader("integer :: l1")
+    fparser2specification = Specification_Part.match(reader)[0][0]
```
Can you just call the constructor rather than having to go via the match method?

arporter replied:
I think so. You can do:

    fparser2spec = Specification_Part(reader).content[0]

sergisiso replied:
Also I didn't get your first comment here, indeed the proposed modification using the constructor looks nicer. I made the change throughout the new unit tests.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
+        :type name: string
+        :param datatype: Data type of the symbol.
+        :type datatype: string
+        :param dimensions: Dimensions of the symbol (0 represents an scalar
```
So `dimensions` can be a list of int or 0? It might be a bit nicer to make sure it's always a list, albeit empty if it's a scalar.

sergisiso replied:
Now is always a list, and I renamed it to 'shape' to make it clear. Scalars are an empty list.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2617,8 +2627,269 @@ def test_binaryoperation_can_be_printed():
+    assert fake_parent.symbol_table.lookup("arg1").name == "arg1"
+    assert fake_parent.symbol_table.lookup("arg1").datatype == 'integer'
+    assert fake_parent.symbol_table.lookup("arg1").dimensions == 0
+    assert fake_parent.symbol_table.lookup("arg1").kind == 'read_arg'
```
Seeing this used in context, I think `kind` is not a good name (it will certainly confuse Fortran programmers like me!). How about `access` instead?

sergisiso replied:
I used a generic 'kind' name as I couldn't think of a better name, but I like 'access' more. Done.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2617,8 +2627,269 @@ def test_binaryoperation_can_be_printed():
+    assert fake_parent.symbol_table.lookup("array3").dimensions == 1
+    assert fake_parent.symbol_table.lookup("array3").kind == 'read_arg'
+
+    reader = FortranStringReader("integer, dimension(3,5) :: array4")
```
In this case do we lose the information that the array is (3,5)? I thought that the `dimensions` component of the symbol would be `[3, 5]` in this case.

sergisiso replied:
Now it is. (in the 'shape' attribute)

arporter replied:
Great.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2617,8 +2627,269 @@ def test_binaryoperation_can_be_printed():
+    assert fake_parent.symbol_table.lookup("arg1").datatype == 'integer'
+    assert fake_parent.symbol_table.lookup("arg1").dimensions == 0
+    assert fake_parent.symbol_table.lookup("arg1").kind == 'read_arg'
+
```
Please break-out the following tests (for differing values of intent) into a separate test.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2617,8 +2627,269 @@ def test_binaryoperation_can_be_printed():
+    assert fake_parent.symbol_table.lookup("arg4").dimensions == 0
+    assert fake_parent.symbol_table.lookup("arg4").kind == 'rw_arg'
+
+    # Test different array dimensions
```
Please put these array-dimensions tests into a separate test.

sergisiso replied:
Done

sergisiso commented on doc/developers.rst:
```diff
@@ -1354,18 +1354,29 @@ stores the resulting AST in `Kern._fp2_ast` for return by future calls.
 
-When a translation to PSyIRe is needed, an ASTProcessor can be used.
+Alternatively, one can call the `psyclone.psyGen.Kern.get_kernel_schedule()`
+to generate the PSyIRe re/presentation of the kernel code. 
```
Done (in the whole subsection for consistency)

sergisiso commented on doc/developers.rst:
```diff
@@ -1354,18 +1354,29 @@ stores the resulting AST in `Kern._fp2_ast` for return by future calls.
     :members:
 
+The results of `psyclone.psyGen.Kern.get_kernel_schedule` is a
+`psyclone.psyGen.KernelSchedule` which has the same functionality than
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -2991,6 +2991,7 @@ def __init__(self, KernelArguments, call, parent=None, check=True):
         self._module_code = call.ktype._ast
         self._kernel_code = call.ktype.procedure
         self._fp2_ast = None  # The fparser2 AST for the kernel
+        self._kern_schedule = None  # PSyIRe schedule for the kernel
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3002,6 +3003,21 @@ def __init__(self, KernelArguments, call, parent=None, check=True):
+    def get_kernel_schedule(self):
+        '''
+        Returns a PSyIRe Schedule representing the kernel code. The Schedule
+        is just generated on first invocation, this allows to retain
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3963,6 +3979,116 @@ def nodes_to_code_block(parent, statements):
+        '''
+        Create a KernelSchedule from the supplied fparser2 AST.
+
+        :param name: Name of the subroutine representing the kernel.
```
This is more concise, I like it. I will use it from now on.

arporter replied:
Thanks :-)

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3963,6 +3979,116 @@ def nodes_to_code_block(parent, statements):
+
+        :param name: Name of the subroutine representing the kernel.
+        :type name: string
+        :param module_ast: fparser2 AST of the full module where the kernel
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3963,6 +3979,116 @@ def nodes_to_code_block(parent, statements):
+        :param module_ast: fparser2 AST of the full module where the kernel
+                           code is located.
+        :type module_ast: :py:class:`fparser.two.Fortran2003.Program`
+        :raises: InternalError: Unexpected fpaser2 AST format.
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3963,6 +3979,116 @@ def nodes_to_code_block(parent, statements):
+        '''
+        from fparser.two import Fortran2003
+
+        def first_type_match(nodelist, typekind):
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3963,6 +3979,116 @@ def nodes_to_code_block(parent, statements):
+                    return x
+            raise ValueError  # Type not found
+
+        def search_subroutine(nodelist, searchname):
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3963,6 +3979,116 @@ def nodes_to_code_block(parent, statements):
+        Transform the fparser2 AST declarations to symbols into the PSyIRe
+        parent node symbol table.
+
+        :param parent: PSyIRe node to insert the symbols found.
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3963,6 +3979,116 @@ def nodes_to_code_block(parent, statements):
+
+        :param parent: PSyIRe node to insert the symbols found.
+        :type parent: :py:class:`psyclone.psyGen.KernelSchedule`
+        :param nodes: fparser2 AST nodes to search for Declarations Statements
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3963,6 +3979,116 @@ def nodes_to_code_block(parent, statements):
+                elif "intent(out)" in str(attr).lower().replace(' ', ''):
+                    decltype = 'write_arg'
+                elif "intent(inout)" in str(attr).lower().replace(' ', ''):
+                    decltype = 'rw_arg'
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3963,6 +3979,116 @@ def nodes_to_code_block(parent, statements):
+
+            # Get name and insert into the parent symbol table
+            for entity in walk_ast(decl.items, [Fortran2003.Entity_Decl]):
+                (name, array_spec, char_len, initialization) = entity.items
```
Yes, the fparser2 Entity_Decl match method returns this 4-element tuple with a None value in the missing parts.

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
+    of the symbol, its datatype, the number of dimensions and the symbol kind
+    (e.g. local, external, read_arg, write_arg, rw_arg).
+
+    :param name: Name of the symbol.
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
+    :type name: string
+    :param datatype: Data type of the symbol.
+    :type datatype: string
+    :param dimensions: Dimensions of the symbol (0 represents an scalar
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3963,6 +3979,116 @@ def nodes_to_code_block(parent, statements):
+            for entity in walk_ast(decl.items, [Fortran2003.Entity_Decl]):
+                (name, array_spec, char_len, initialization) = entity.items
+                if (array_spec is not None) or (initialization is not None):
+                    raise NotImplementedError()
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
+
+    @property
+    def name(self):
+        return self._name
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
+    '''
+    Encapsulates the symbols table and provides methods to declare new symbols
+    and look up existing symbols. It is implemented as a monolithic scope
+    symbol table.
```
Done?

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
+        :type kind: string
+        '''
+        if name in self._symbols:
+            raise InternalError("Multiple definition of symbol: '{0}'."
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
+
+class SymbolTable(object):
+    '''
+    Encapsulates the symbols table and provides methods to declare new symbols
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
+
+    def declare(self, name, datatype, dimensions, kind):
+        '''
+        Declare a new symbol in the symbols table.
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3963,6 +3979,116 @@ def nodes_to_code_block(parent, statements):
+
+            # Parse number of dimensions if it is an array
+            dimensions = 0
+            for attr in walk_ast(decl.items, [Fortran2003.Assumed_Shape_Spec]):
```
End up refactoring it, I need the loop to keep the order but do different things for each component.

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
+    :param dimensions: Dimensions of the symbol (0 represents an scalar
+                       symbol).
+    :type dimensions: list of integers
+    :param kind: Information about the variable declaration attributes.
```
Updated attribute name and docstrings

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
+    :param kind: Information about the variable declaration attributes.
+    :type kind: string
+    '''
+    def __init__(self, name, datatype=None, dimensions=0, kind=None):
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4178,6 +4304,142 @@ def _number_handler(self, node, parent):
+        :param dimensions: Dimensions of the symbol (0 represents an scalar
+                           symbol).
+        :type dimensions: list of integers
+        :param kind: Information about the variable declaration attributes.
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -3963,6 +3979,116 @@ def nodes_to_code_block(parent, statements):
+                dimensions = dimensions + 1
+
+            # Parse intent attributes
+            decltype = 'local'  # If no intent attribute provided, it is a
```
I added a `specify_argument_list` method (called afterwards) that records the order of the arguments and also corrects the 'intent' if this is not implicit in the declaration.

sergisiso replied:
As discussed, I moved the `specify_argument_list` call inside the `process_declaration`, so the `process_declaration` method returns with a fully initialized Symbol Table. However, I left the method as public and the argument-list as a mandatory argument. (as we always need the arguments to have a Symbol Table with correct attributes, and we can pass an empty list if there aren't arguments)

sergisiso commented:
@arporter, I made a few changes addressing your points:
- Renamed symbol attributes dimensions -> shape and kind -> access.
- Shape is now again a list of dimensions where an empty list represents a scalar and a None value represents an unknown size.
- I added a `specify_argument_list` method that records the order of the arguments a also corrects the 'access' value if needed.
- Also I improved the docstrings, added more tests and brought up-to-date with master.

Back to you for the next review.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3963,6 +3979,116 @@ def nodes_to_code_block(parent, statements):
+        new_schedule = KernelSchedule(name)
+
+        try:
+            # Assume just 1 Fortran module definition in the file
```
Thanks for improving. I don't think this should be an InternalError as it is quite possible for a user to supply us with a Fortran file that will trigger this error. We reserve InternalError for "impossible" cases that should never happen unless there's a bug in PSyclone. Perhaps use GenerationError instead? Also, please could you include the name of the subroutine and/or the module name in the error message.

sergisiso replied:
Done

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4307,14 +4327,23 @@ def search_subroutine(nodelist, searchname):
             raise InternalError("Unexpected kernel AST. Could not find "
                                 "subroutine: {0}".format(name))
 
+        arg_list = []
```
Please add a comment here, e.g. 'save list of kernel arguments'

sergisiso replied:
Done

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4307,14 +4327,23 @@ def search_subroutine(nodelist, searchname):
+            raise InternalError("Unexpected kernel AST. The argument list of"
+                                " '{0}' do not match the variable "
+                                "declarations.".format(name))
+
```
Good error message. Please could you tweak it slightly, e.g. "The kernel argument list '{0}' does not match the .... for kernel XXXX"

sergisiso replied:
Done

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4327,12 +4356,12 @@ def search_subroutine(nodelist, searchname):
     def process_declarations(self, parent, nodes):
         '''
-        Transform the fparser2 AST declarations to symbols into the PSyIRe
+        Transform the fparser2 AST declarations to symbols into the PSyIR
```
"Transform the variable declarations in the fparser2 parser tree into symbols in the ...."

sergisiso replied:
Done

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4346,30 +4375,53 @@ def process_declarations(self, parent, nodes):
+                                  Fortran2003.Assumed_Size_Spec,
+                                  ]):
+                if isinstance(attr, Fortran2003.Assumed_Size_Spec):
+                    raise InternalError("Assumed size arrays are not "
```
Great.

sergisiso replied:
Sure

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4590,81 +4642,142 @@ def _number_handler(self, node, parent):
-                                "".format(name))
-        else:
-            self._symbols[name] = Symbol(name, datatype, dimensions, kind)
+            raise InternalError("Symbol table already contains a symbol with"
```
Thanks for improving the error message. On reflection I don't think this should be an InternalError but I'm not sure what it should be. Maybe ValueError or KeyError?

sergisiso replied:
I used KeyError

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4590,81 +4642,142 @@ def _number_handler(self, node, parent):
+                                " name '{0}'.".format(name))
+        self._symbols[name] = Symbol(name, datatype, shape, access)
+
+    def specify_argument_list(self, argument_name_list):
```
As suggested earlier, I think we can remove this method.

sergisiso replied:
I still need to keep it to keep track of the argument order, but now is called from `process_declarations`

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4465,6 +4639,201 @@ def _number_handler(self, node, parent):
+    table to keep a record of the declared variables and their attributes.
+
+    :param str name: Kernel subroutine name
+    :type name: string
```
You don't need this `:type ...` line.

sergisiso replied:
Deleted

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4465,6 +4639,201 @@ def _number_handler(self, node, parent):
+        self._symbol_table = SymbolTable()
+
+    @property
+    def symbol_table(self):
```
Please document this method (i.e. the return type).

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2621,8 +2631,465 @@ def test_binaryoperation_can_be_printed():
+def test_symboltable_declare():
+    '''Test that the declare method inserts new symbols in the symbol
+    table, but raises appropiate errors when provied with wrong parameters
+    or duplate declarations'''
```
s/duplate/duplicate/ and full stop please.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2621,8 +2631,465 @@ def test_binaryoperation_can_be_printed():
+
+
+def test_symboltable_lookup():
+    '''Test that the lookup method retrive symbols from the symbol table
```
s/retrive/retrieves/

arporter replied:
and, below, "exists, otherwise it raises an error."

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2621,8 +2631,465 @@ def test_binaryoperation_can_be_printed():
+
+def test_fparser2astprocessor_generate_schedule_two_modules():
+    ''' Tests the fparser2AST generate_schedule method with an simple
+    subroutine.
```
Please update this docstring to say that we're testing that we reject files containing more than one module definition.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2621,8 +2631,465 @@ def test_binaryoperation_can_be_printed():
+    ''' Tests the fparser2AST generate_schedule method with an simple
+    subroutine.
+    '''
+    ast1 = fpapi.parse(FAKE_KERNEL_METADATA*2, ignore_comments=True)
```
That's clever!

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2621,8 +2631,465 @@ def test_binaryoperation_can_be_printed():
+    ast2 = my_kern.ast
+    processor = Fparser2ASTProcessor()
+
+    # Test properly formed kernel module
```
Please update this comment.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2621,8 +2631,465 @@ def test_binaryoperation_can_be_printed():
+
+
+def test_fparser2astprocessor_generate_schedule_dummy_subroutine():
+    ''' Tests the fparser2AST generate_schedule method with an simple
```
s/an simple/a simple/

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2621,8 +2631,465 @@ def test_binaryoperation_can_be_printed():
+    schedule = processor.generate_schedule("dummy_code", ast2)
+    assert isinstance(schedule, KernelSchedule)
+
+    # Test argument intent is infered when not available in the declaration
```
s/infered/inferred/

sergisiso replied:
I will need to enable vim spellchecker :0. Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2621,8 +2631,465 @@ def test_binaryoperation_can_be_printed():
+
+
+def test_fparser2astprocessor_generate_schedule_unmatching_arguments():
+    ''' Tests the fparser2AST generate_schedule method with an simple
```
Please correct this docstring to describe the test.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2621,8 +2631,465 @@ def test_binaryoperation_can_be_printed():
+    ''' Tests the fparser2AST generate_schedule method with an simple
+    subroutine.
+    '''
+    DUMMY_KERNEL_METADATA = '''
```
This name should probably be lower case as the variable has local scope (not module). Same applies to the earlier one as well.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2621,8 +2631,465 @@ def test_binaryoperation_can_be_printed():
+    fparser2specification = Specification_Part.match(reader)[0][0]
+    with pytest.raises(InternalError) as error:
+        processor.process_declarations(fake_parent, [fparser2specification])
+    assert ("Array initializations on the declaration statements are not "
```
Strictly speaking this is not an array initialisation. Please could you correct the error message?

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2621,8 +2631,465 @@ def test_binaryoperation_can_be_printed():
+            "supported.") in str(error.value)
+
+    # Char lengths are not supported
+    # TODO: It would be simpler to do just a Specification_Part.match() instead
```
SpecificationPart(reader) without the match()

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2621,8 +2631,465 @@ def test_binaryoperation_can_be_printed():
+    from fparser.two.parser import ParserFactory
+    from fparser.common.readfortran import FortranStringReader
+    from fparser.two.Fortran2003 import Specification_Part
+    f2008parser = ParserFactory().create(std="f2008")
```
Rather than create a new parser for each test (that needs it) in this module, please add a fixture at the beginning of the module that just creates one on module entry.

    @pytest.fixture(scope="module")
    def f2008_parser():
        return ParserFactory().create(std="f2008")

you then provide this fixture to each test that needs it (as an argument to the test method).

sergisiso replied:
Good one

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2621,8 +2631,465 @@ def test_binaryoperation_can_be_printed():
+
+
+def test_fparser2astprocessor_process_declarations_array_attributes():
+    '''Test that process_declarations method parses multiple specifications
```
"handles various different specifications of array attributes."?

sergisiso replied:
Updated, but with s/array/variable/

arporter commented:
Back to @sergisiso.

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4346,30 +4375,53 @@ def process_declarations(self, parent, nodes):
-                dimensions = dimensions + 1
+                                 [Fortran2003.Assumed_Shape_Spec,
+                                  Fortran2003.Explicit_Shape_Spec,
+                                  Fortran2003.Assumed_Size_Spec,
```
Removed now

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4346,30 +4375,53 @@ def process_declarations(self, parent, nodes):
+                                  Fortran2003.Int_Literal_Constant):
+                        shape.append(int(attr.items[1].items[0]))
+                    else:
+                        raise InternalError("Only integer literals are "
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4346,30 +4375,53 @@ def process_declarations(self, parent, nodes):
-            # local variable.
+            decltype = 'local'  # If no intent attribute is provided, it is
+            # provisionally marked as a local variable (when the argument
+            # list is parsed, undeclared arguments are updated appropriately).
```
Yes. Fixed.

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4346,30 +4375,53 @@ def process_declarations(self, parent, nodes):
-                    raise NotImplementedError()
-                parent.symbol_table.declare(str(name), datatype, dimensions,
+                if (array_spec is not None):
+                    raise InternalError("Array specifications after the "
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4590,81 +4642,142 @@ def _number_handler(self, node, parent):
+    :param str datatype: Data type of the symbol.
+    :param int shape: Shape of the symbol (an empty list represents \
+                      a scalar symbol).
+    :param list access: Information to specify if the declaration represents
```
Check the new description. Is there a better way in the docstring to specify it is a list of str?

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4590,81 +4642,142 @@ def _number_handler(self, node, parent):
+    def shape(self):
+        '''
+        :return: Shape (number of dimensions and sizes) of the Symbol.
+        :rtype: integer
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4590,81 +4642,142 @@ def _number_handler(self, node, parent):
+    def access(self):
+        '''
+        :return: Whether the variable is local, external or an argument. In
+                 case it is argument it also specify if it is a read, write
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4590,81 +4642,142 @@ def _number_handler(self, node, parent):
+            raise InternalError("Symbol shape list elements can only be "
+                                "'integer' or 'None'.")
+
+        if access not in ('local', 'external', 'read_arg', 'write_arg',
```
By your example I believe you meant a class variable, and I used a tuple instead of a list to make it a bit more immutable. Also I extended the docstring with a definition of each access type.

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4590,81 +4642,142 @@ def _number_handler(self, node, parent):
+    @access.setter
+    def access(self, new_access):
+        '''
+        :param str new_access: New value of the access attribute, can be:
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4590,81 +4642,142 @@ def _number_handler(self, node, parent):
+    def __init__(self, name, datatype=None, shape=[], access=None):
+
+        if datatype not in ('real', 'integer'):
+            raise InternalError("Symbol can only be initialized with 'real' "
```
Updated this and the other error types in Symbol.__init__

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4590,81 +4642,142 @@ def _number_handler(self, node, parent):
+                      symbol).
+        :type shape: list of integers
+        :param str access: Information about the variable declaration
+                           attributes.
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4590,81 +4642,142 @@ def _number_handler(self, node, parent):
-        :type kind: string
+        :param str name: Name of the symbol.
+        :param str datatype: Datatype of the symbol.
+        :param shape: Shape of the symbol (empty list represents an scalar
```
OK, I extended the Symbol class and methods docstrings and started a PSyIR chapter in the Developer's Guide. The chapter is in second place to precede all other mentions of PSyIR on the guide. I made some changes in the Node docstrings (PSyIRe->PSyIR, capitalization and punctuations, ... ) to make the text more coherent.

sergisiso commented:
Back to you @arporter , I addressed your comments, the main changes are:
- New chapter in the developer's guide for PSyIR.
- More meaningful error types and messages.
- Now the argument_list is passed to the process_declaration method, so it is self contained and it returns with a Symbol Table populated with appropriate values.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4250,6 +4266,188 @@ def nodes_to_code_block(parent, statements):
+            if str(decl.items[0]).lower().startswith('real'):
+                datatype = 'real'
+            elif str(decl.items[0]).lower().startswith('integer'):
+                datatype = 'integer'
```
Please add `else NotImplementedError` so that we don't silently ignore things that we don't understand.

sergisiso replied:
Done

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4250,6 +4266,188 @@ def nodes_to_code_block(parent, statements):
+                                                  "Only integer literals are "
+                                                  "supported for explicit "
+                                                  "shape array declarations."
+                                                  "".format(nodes))
```
Please add an `else InternalError` to defend ourselves against a future developer modifying the call to `walk_ast` but then forgetting to update this if block. I expect you'll then have to monkeypatch `walk_ast` in order to get a test to cover this branch of the if block.

sergisiso replied:
Done, I have moved the parsing of dimension attributes to a static private method, so it is easy to monkeypatch just this specific `walk_ast`.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4465,6 +4663,227 @@ def _number_handler(self, node, parent):
+    Symbol item for the Symbol Table. It contains information about: the name,
+    the datatype, the shape (in row-major order) and the symbol
+    access characterisctics. The symbol access can be:
+        - local: Variable that just exist in the kernel scope.
```
Great documentation. Please put the possible symbol access values in quotes to show that they are strings, i.e. "local": Variable that...

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4465,6 +4663,227 @@ def _number_handler(self, node, parent):
+                       index is contiguous in memory). Each entry represents \
+                       an array dimension. If not None then it holds the \
+                       extent of that dimension. If it is an empy list it \
+                       represents an scalar.
```
"...empty list then the symbol represents a scalar." Please also explain that one or more entries of None means that it is an assumed-shape array. I'm not sure what this means in a language-independent context - maybe it's just a pointer?

sergisiso replied:
Modified sentence and clarified that the extent is unknown (from PSyIR point of view). 'assumed-shape' is Fortran-specific and 'pointer' is probably too implementation-specific.

arporter replied:
Great.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4465,6 +4663,227 @@ def _number_handler(self, node, parent):
+                            "'integer' or 'None'.")
+
+        if access not in Symbol.valid_access_types:
+            raise ValueError("Symbol access attribute can only be: "
```
You can use `Symbol.valid_access_types` in this error message to future-proof it against any changes to the valid access types, e.g.:

    "...can only be one of '{0}' but got '{1}'".format(str(Symbol.valid_access_types), access)

sergisiso replied:
Neat

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4465,6 +4663,227 @@ def _number_handler(self, node, parent):
+        :raises ValueError: New access parameter has an invalid value.
+        '''
+        if new_access not in Symbol.valid_access_types:
+            raise ValueError("Symbol access attribute can only be: "
```
Please use `Symbol.valid_access_types` in the error message for future proofing.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2621,17 +2638,508 @@ def test_binaryoperation_can_be_printed():
+    # Test argument intent is inferred when not available in the declaration
+    assert schedule.symbol_table.lookup('f3').access == 'readwrite_arg'
+
+    # Test argument intent is inferred when not available in the declaration
```
This comment needs updating?

sergisiso replied:
Ups, fixed!

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2621,17 +2638,508 @@ def test_binaryoperation_can_be_printed():
+    # Test kernel with no arguments, should still proceed
+    schedule = processor.generate_schedule("dummy_code", ast2)
+    assert isinstance(schedule, KernelSchedule)
+    # TODO: In the future we could validate that metadata matches
```
Great. I was just thinking "but that kernel is not valid" when I got to this comment :-)

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2621,17 +2638,508 @@ def test_binaryoperation_can_be_printed():
+
+def test_fparser2astprocessor_process_declarations(f2008_parser):
+    '''Test that process_declarations method of fparse2astprocessor
+    converts the fparser2 decalations to symbols in the provided
```
s/decalations/declarations/

sergisiso replied:
Fixed

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2621,17 +2638,508 @@ def test_binaryoperation_can_be_printed():
+
+
+def test_fparser2astprocessor_handling_assignment_stmt(f2008_parser):
     ''' Test that fparser2 Assignment_Stmt is converted to expected PSyIRe
```
Although you've not strictly touched it in this Issue, please could you do s/PSyIRe/PSyIR/ in the docstrings of these tests.

sergisiso replied:
Done

arporter commented:
Back to @sergisiso.

sergisiso commented on doc/developers.rst:
```diff
@@ -1426,18 +1458,29 @@ stores the resulting AST in `Kern._fp2_ast` for return by future calls.
+`psyclone.psyGen.KernelSchedule` which has the same functionality as
+a PSy Schedule but with the addition of a Symbol Table.
+
+.. autoclass:: psyclone.psyGen.SymbolTable
```
Removed autoclass and added a reference.

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4250,6 +4266,188 @@ def nodes_to_code_block(parent, statements):
+            for x in nodelist:
+                if (isinstance(x, Fortran2003.Subroutine_Subprogram) and
+                   str(x.content[0].get_name()) == searchname):
+                        return x
```
Fixed

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4250,6 +4266,188 @@ def nodes_to_code_block(parent, statements):
+                        return x
+            raise ValueError  # Subroutine not found
+
+        new_schedule = KernelSchedule(name)
```
Could you clarify? searchname is just used in the local function that  ends with the `raise ValueError` line

arporter replied:
My mistake - I failed to spot that `search_subroutine` was contained within the routine.

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4250,6 +4266,188 @@ def nodes_to_code_block(parent, statements):
+                elif "intent(out)" in str(attr).lower().replace(' ', ''):
+                    decltype = 'write_arg'
+                elif "intent(inout)" in str(attr).lower().replace(' ', ''):
+                    decltype = 'readwrite_arg'
```
Added a `else NotImplementedError` instead, as there are some valid Fortran attributes that are not recognized yet. (note I also refactored part of this method and loop)

sergisiso commented:
Back to you @arporter . This time I mainly refactored the process_declaration method to improve the exceptions for unsupported/unexpected declaration parsing. Also addressed your other comments.

arporter commented:
* All tests pass (inc. examples) and all new code is covered.
* No x-failing tests are affected.
* Documentation has been updated.
* Branch is up-to-date with master.

arporter commented:
* Big picture is up-to-date.
* Documentation builds OK.
* This is 'internal' functionality so no need for a new example.
* LFRic functionality is not affected.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4490,6 +4506,252 @@ def nodes_to_code_block(parent, statements):
+            :param list nodelist: List of fparser2 nodes.
+            :param type typekind: The fparse2 Type we are searching for.
+            '''
+            for x in nodelist:
```
pylint complains about using just `x` as a variable name.

sergisiso replied:
Fair enough

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4490,6 +4506,252 @@ def nodes_to_code_block(parent, statements):
+                                      "module definition per file supported."
+                                      "".format(name))
+            mod_content = module_ast.content[0].content
+            mod_spec = first_type_match(mod_content,
```
pylint complains that `mod_spec` is unused. Perhaps just do `_ = first_type_match(...)`

sergisiso replied:
I am not using this at the moment, it is to access the metadata, I commented it and added a TODO referencing #288

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4490,6 +4506,252 @@ def nodes_to_code_block(parent, statements):
+            '''
+            if nodes is None:
+                return []
+            else:
```
pylint points out that this `else` is unnecessary because it follows a `return`

sergisiso replied:
Fixed

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4490,6 +4506,252 @@ def nodes_to_code_block(parent, statements):
+            else:
+                if type(nodes).__name__.endswith("_List"):
+                    return nodes.items
+                else:
```
Again, this `else` is unnecessary.

sergisiso replied:
Fixed

sergisiso commented:
What do you think of this new descriptions  @rupertford @arporter ?
- read_arg: It is defined at the entry of the kernel and remains constant.
- write_arg: It is not be defined at the entry of the kenrel and it will be used to return data to the kernel caller.
- readwrite_arg: It is defined at the entry of the kernel and it will be used to return data to the kernel caller.

arporter commented:
You 'just' need to update the description of the access patterns for the Symbol Table and fix a few pylint/pycodestyle errors and then I'm happy (and hopefully so is @rupertford ;-))

sergisiso commented:
Done, back to you @arporter (and @rupertford to comment on SymbolTable 'access' docstring  )

rupertford commented:
> What do you think of this new descriptions @rupertford @arporter ?
> 
> * read_arg: It is initialised on entry to the kernel and remains constant.
> * write_arg: It is not initialised on entry to the kernel (and therefore it needs to be written before it can be read) and it will be used to return data to the kernel caller.
> * readwrite_arg: It is initialised on entry to the kernel and it will be used to return data to the kernel caller.

It depends on what information you are trying to present. The new description explains whether data might need to be passed into and out-of a kernel given the access pattern but does not explain what the access pattern actually is. Assuming you are talking about kernels, rather than regions of code, you could say something like ...

read_arg: the argument is only read within the kernel. Therefore the arguments data only needs to be passed into the kernel.
write_arg: the first access to the argument is a write within the kernel before any optional reads or subsequent writes. Therefore the arguments data only needs to be passed out of the kernel.
read_write_arg: the first access to the argument is a read within the kernel and this is followed by optional reads and at least one write. Therefore the arguments data must be passed into and out-of the kernel.

Presumably we are not worrying about accessing parts of an argument here e.g. modifying halo data rather than owned data?

sergisiso commented:
@arporter it is ready for review again. I added the Symbol and SymbolTable modifications we discussed yesterday. Finally I decided for local-scoped symbols to have the input/output always False instead of None, this is semantically true and the handling will be easier. Also there is just global_arguments in the global_* category at the moment, the other global declarations in the specification part should raise an exception.  If we add more globals in the future we could subclass them instead of just using a postfix.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4705,6 +4968,300 @@ def _number_handler(self, node, parent):
+    @is_output.setter
+    def is_output(self, new_is_output):
+        '''
+        :param bool new_is_output: Whether it is survives outside the kernel \
```
"Whether the variable represented by this symbol survives outside..."

sergisiso replied:
Fixed

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4705,6 +4968,300 @@ def _number_handler(self, node, parent):
+
+    def specify_argument_list(self, argument_name_list):
+        '''
+        Keep track of the arguments order and provide the scope, is_input and
```
"track of the order of the arguments and..."

sergisiso replied:
Done

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4705,6 +4968,300 @@ def _number_handler(self, node, parent):
+    @property
+    def symbol_table(self):
+        '''
+        :return: Table containing symbol information of the kernel.
```
s/of the/for the/

sergisiso replied:
Fixed

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -54,7 +54,8 @@
     OMPParallelDirective, OMPDoDirective, OMPDirective, Directive, CodeBlock, \
-    Assignment, Reference, BinaryOperation, Array, Literal, Node, IfBlock
+    Assignment, Reference, BinaryOperation, Array, Literal, Node, IfBlock, \
+    BinaryOperation, KernelSchedule, Symbol, SymbolTable
```
You have `BinaryOperation` twice in this list :-)

sergisiso replied:
Fixed

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -70,6 +71,14 @@
+# Module fixtures
+
+@pytest.fixture(scope="module")
+def f2008_parser():
```
Please add a docstring describing this fixture.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2668,17 +2692,609 @@ def test_binaryoperation_can_be_printed():
+    assert isinstance(sym_table.lookup("var3"), Symbol)
+    assert sym_table.lookup("var3").name == "var3"
+
+    with pytest.raises(KeyError) as error:
```
pylint points out that you don't use `error`

sergisiso replied:
Fixed

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2668,17 +2692,609 @@ def test_binaryoperation_can_be_printed():
+
+    # Test kernel with two modules
+    with pytest.raises(GenerationError) as error:
+        schedule = processor.generate_schedule("dummy_code", ast2)
```
pylint complains that `schedule` is not used (just do `_ = processor.generate...`)

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2668,17 +2692,609 @@ def test_binaryoperation_can_be_printed():
+    del ast2.content[0].content[2].content[1].content[2]
+    schedule = processor.generate_schedule("dummy_code", ast2)
+    assert isinstance(schedule, KernelSchedule)
+    assert len(schedule.children) == 0
```
This should be `assert not schedule.children`

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2668,17 +2692,609 @@ def test_binaryoperation_can_be_printed():
+
+    # Test exception for unmatching argument list
+    with pytest.raises(InternalError) as error:
+        schedule = processor.generate_schedule("dummy_code", ast2)
```
`schedule` is not used.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2668,17 +2692,609 @@ def test_binaryoperation_can_be_printed():
+            "supported.") in str(error.value)
+
+
+def test_fparser2astprocessor_process_not_supported_declarations(f2008_parser):
```
This test routine does not use `f2008_parser`

sergisiso replied:
But they still need the ParserFactory to create a parser right? If I call the test in isolation without the fixture it fails.

arporter replied:
Oops, yes, you are quite right!

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2668,17 +2692,609 @@ def test_binaryoperation_can_be_printed():
+    assert ". Unrecognized attribute " in str(error.value)
+
+
+def test_fparser2astprocessor_process_declarations_intent(f2008_parser):
```
This one also doesn't use `f2008_parser`

sergisiso replied:
Same as above.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2668,17 +2692,609 @@ def test_binaryoperation_can_be_printed():
+
+
+def test_fparser2astprocessor_parse_array_dimensions_attributes(
+        f2008_parser):
```
and again.

sergisiso replied:
Same as above.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2668,17 +2692,609 @@ def test_binaryoperation_can_be_printed():
+    from fparser.two.Fortran2003 import Dimension_Attr_Spec
+    import fparser
+
+    def walk_ast_return(arg1, arg2):
```
Please add a very brief docstring saying what this routine is for.

arporter replied:
and please do `class invalid(object)` to make pylint happier.

sergisiso replied:
Done

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4705,6 +4968,300 @@ def _number_handler(self, node, parent):
+                      mechanism, at the moment just 'global_argument' is \
+                      available for variables passed in/out of the kernel \
+                      by argument.
+    :param bool is_input: Whether it is an already existent data that is \
```
"Whether the symbol represents data that exists before the kernel is entered and that is passed into the kernel." (and elsewhere please)

sergisiso replied:
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -5115,29 +5170,41 @@ def declare(self, name, datatype, shape, access):
             symbol = self.lookup(name)
-            if symbol.access == 'local':
-                symbol.access = 'readwrite_arg'
+            if symbol.scope == 'local':
```
Done

sergisiso commented on src/psyclone/psyGen.py:
```diff
@@ -4490,6 +4506,253 @@ def nodes_to_code_block(parent, statements):
+
+    def process_declarations(self, parent, nodes, arg_list):
+        '''
+        Transform the variable declarations in fparser2 parse tree into
```
Done

sergisiso commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2810,9 +2816,9 @@ def test_symboltable_lookup():
     '''Test that the lookup method retrives symbols from the symbol table
     if the name exists, otherwise it raises and error.'''
```
Fixed

sergisiso commented:
Back to you @arporter . I made most of the requested changes (the exception is the f2008_parser fixture which I think is still needed.)
